frame/support/src/storage/types/mod.rs

```diff
@@ -0,0 +1,100 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Storage types to build abstraction on storage, they implements storage traits such as
+//! StorageMap and others.
+
+use codec::FullCodec;
+use frame_metadata::{DefaultByte, StorageEntryModifier};
+
+mod value;
+mod map;
+mod double_map;
+
+pub use value::{StorageValue, StorageValueMetadata};
+pub use map::{StorageMap, StorageMapMetadata};
+pub use double_map::{StorageDoubleMap, StorageDoubleMapMetadata};
+
+/// Trait implementing how the storage optional value is converted into the queried type.
+/// It is implemented by:
+/// * `OptionQuery` which convert an optional value to an optional value, user when querying
+///   storage will get an optional value.
+/// * `ValueQuery` which convert an optional value to a value, user when querying storage will get
+///   a value.
+pub trait QueryKindTrait<Value> {
+	/// Metadata for the storage kind.
+	const METADATA: StorageEntryModifier;
+
+	/// Type returned on query
+	type Query: FullCodec + 'static;
+
+	/// Convert an optional value (i.e. some if trie contains the value or none otherwise) to the
+	/// query.
+	fn from_optional_value_to_query<OnEmpty>(v: Option<Value>) -> Self::Query where
+		OnEmpty: crate::traits::Get<Self::Query>;
+
+	/// Convert a query to an optional value.
+	fn from_query_to_optional_value(v: Self::Query) -> Option<Value>;
+}
+
+/// Implement QueryKindTrait with query being `Option<Value>`
+pub struct OptionQuery;
+impl<Value: FullCodec + 'static> QueryKindTrait<Value> for OptionQuery where
+{
+	const METADATA: StorageEntryModifier = StorageEntryModifier::Optional;
+	type Query = Option<Value>;
+	fn from_optional_value_to_query<OnEmpty>(v: Option<Value>) -> Self::Query where
+		OnEmpty: crate::traits::Get<Self::Query>
+	{
+		if v.is_none() {
+			OnEmpty::get()
+		} else {
+			v
+		}
```

Comment:
```suggestion
	v
```

I like your idea that we should document this behavior.

In the end people will probably not use this stuff without the macros anyway.

Reply:
ah my [comment](https://github.com/paritytech/substrate/pull/7278#discussion_r503799789) is partially wrong 
decl_storage currently allows `MyStorage: Option<u32> = Some(3);` and it returns `Some(3)` when value doesn't exist.

So if we don't want to break any storage during the migration we should keep the if condition here.

Also at the end it is not so so bad, it is indeed useless (as nobody should use this feature), but on the implementation can handle it. Thus I'm mixed

Reply:
I found an alternative: make compilation fail when a OnEmpty different from GetDefault is used alongside OptionQuery:
```rust
		type SomeInvalidExample = StorageValue<Prefix, u32, OptionQuery, ADefault>;
		SomeInvalidExample::get();
```
Now fails with 
```
error[E0599]: no function or associated item named `get` found for struct `types::value::StorageValue<types::value::test::Prefix, u32, types::OptionQuery, types::value::test::ADefault>` in the current scope
   --> frame/support/src/storage/types/value.rs:224:23
    |
36  | / pub struct StorageValue<Prefix, Value, QueryKind=OptionQuery, OnEmpty=GetDefault>(
37  | |     core::marker::PhantomData<(Prefix, Value, QueryKind, OnEmpty)>
38  | | );
    | |__- function or associated item `get` not found for this
...
224 |           SomeInvalidExample::get();
    |                               ^^^ function or associated item not found in `types::value::StorageValue<types::value::test::Prefix, u32, types::OptionQuery, types::value::test::ADefault>`
    | 
   ::: frame/support/src/storage/types/mod.rs:59:1
    |
59  |   pub struct OptionQuery;
    |   ----------------------- doesn't satisfy `_: QueryKindTrait<u32, types::value::test::ADefault>`
    |
    = note: the method `get` exists but the following trait bounds were not satisfied:
            `types::OptionQuery: QueryKindTrait<u32, types::value::test::ADefault>`

```

I think we should either we do this or support a not-none on empty for OptionQuery. I like not to allow this weird usage as well.